lib/puppet_x/puppetlabs/scheduled_task/taskscheduler2.rb

```diff
@@ -0,0 +1,318 @@
+      definition.Principal.LogonType = TASK_LOGON_PASSWORD
+      definition.Principal.RunLevel = TASK_RUNLEVEL_HIGHEST
+      return true
+    end
+  end
+
+  # Returns the compatibility level of the task.
+  #
+  def self.compatibility(definition)
+    definition.Settings.Compatibility
+  end
+
+  # Sets the compatibility with the task.
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa381846(v=vs.85).aspx
+  #
+  def self.set_compatibility(definition, value)
+    definition.Settings.Compatibility = value
+  end
+
+  # Task Actions
+  # Returns the number of actions associated with the active task.
+  #
+  def self.action_count(definition)
+    definition.Actions.count
+  end
+
+  def self.action(definition, index)
+    result = nil
+
+    begin
+      result = definition.Actions.Item(index)
+    rescue WIN32OLERuntimeError => err
+      # E_INVALIDARG 0x80070057 from # https://msdn.microsoft.com/en-us/library/windows/desktop/aa378137%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396
+      if err.message =~ /80070057/m
+        result = nil
+      else
+        raise
+      end
+    end
+
+    result
+  end
+
+  def self.create_action(definition, action_type)
+    definition.Actions.Create(action_type)
+  end
+
+  # Task Triggers
+  def self.trigger_count(definition)
+    definition.Triggers.count
+  end
+
+  # Returns a Win32OLE Trigger Object for the trigger at the given index for the
+  # supplied definition.
+  #
+  # Returns nil if the index does not exist
+  #
+  # Note - This is a 1 based array (not zero)
+  #
+  def self.trigger(definition, index)
+    result = nil
+
+    begin
+      result = definition.Triggers.Item(index)
+    rescue WIN32OLERuntimeError => err
+      # E_INVALIDARG 0x80070057 from # https://msdn.microsoft.com/en-us/library/windows/desktop/aa378137%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396
+      if err.message =~ /80070057/m
+        result = nil
+      else
+        raise
+      end
+    end
+
+    result
+  end
+
+  def self.append_new_trigger(definition, trigger_type)
+    definition.Triggers.create(trigger_type)
+  end
+
+  # Deletes the trigger at the specified index.
+  #
+  def self.delete_trigger(definition, index)
+    definition.Triggers.Remove(index)
+
+    index
+  end
+
+  # Helpers
+
+  # From https://msdn.microsoft.com/en-us/library/windows/desktop/aa381850(v=vs.85).aspx
+  # https://en.wikipedia.org/wiki/ISO_8601#Durations
+  #
+  # The format for this string is PnYnMnDTnHnMnS, where nY is the number of years, nM is the number of months,
+  # nD is the number of days, 'T' is the date/time separator, nH is the number of hours, nM is the number of minutes,
+  # and nS is the number of seconds (for example, PT5M specifies 5 minutes and P1M4DT2H5M specifies one month,
+  # four days, two hours, and five minutes)
+  def self.duration_to_hash(duration)
+    regex = /^P((?'year'\d+)Y)?((?'month'\d+)M)?((?'day'\d+)D)?(T((?'hour'\d+)H)?((?'minute'\d+)M)?((?'second'\d+)S)?)?$/
```

Comment:
So I suspect we don't have to worry about this because we always get back the format that we want, with everything specified... but... it does look like something like `PT36H` is valid... which will cause some of the matches to be `nil`

```
[5] pry(main)> regex = /^P((?'year'\d+)Y)?((?'month'\d+)M)?((?'day'\d+)D)?(T((?'hour'\d+)H)?((?'minute'\d+)M)?((?'second'\d+)S)?)?$/
=> /^P((?'year'\d+)Y)?((?'month'\d+)M)?((?'day'\d+)D)?(T((?'hour'\d+)H)?((?'minute'\d+)M)?((?'second'\d+)S)?)?$/
[6] pry(main)> matches = regex.match('PT36H')
=> #<MatchData "PT36H" year:nil month:nil day:nil hour:"36" minute:nil second:nil>
[7] pry(main)> matches['month']
=> nil
```

I didn't see any problems around this when testing so I *think* we're OK?

Reply:
As I'm converting this to a hash, the missing `months` would be a nil anyway e.g.

` xxxxx.duration_to_hash('PT36H')[:months]
=> nil
`

So either way, they'll be nil.

Currently the V1 API uses milliseconds not the duration format so the tests are safe.  Perhaps when we use the V2 API things may be tripped up e.g.

'PT2M' == 'PT120S'  (2 minutes equals 120 seconds)